common/network-common/src/main/java/org/apache/spark/network/util/TransportFrameDecoder.java

```diff
@@ -123,31 +139,54 @@ private long decodeFrameSize() {
 
   private ByteBuf decodeNext() {
     long frameSize = decodeFrameSize();
-    if (frameSize == UNKNOWN_FRAME_SIZE || totalSize < frameSize) {
+    if (frameSize == UNKNOWN_FRAME_SIZE) {
       return null;
     }
 
-    // Reset size for next frame.
-    nextFrameSize = UNKNOWN_FRAME_SIZE;
-
-    Preconditions.checkArgument(frameSize < MAX_FRAME_SIZE, "Too large frame: %s", frameSize);
-    Preconditions.checkArgument(frameSize > 0, "Frame length should be positive: %s", frameSize);
-
-    // If the first buffer holds the entire frame, return it.
-    int remaining = (int) frameSize;
-    if (buffers.getFirst().readableBytes() >= remaining) {
-      return nextBufferForFrame(remaining);
+    if (frameBuf == null) {
+      Preconditions.checkArgument(frameSize < MAX_FRAME_SIZE,
+          "Too large frame: %s", frameSize);
+      Preconditions.checkArgument(frameSize > 0,
+          "Frame length should be positive: %s", frameSize);
+      frameRemainingBytes = (int) frameSize;
+
+      // If buffers is empty, then return immediately for more input data. Otherwise, if the
+      // first buffer holds the entire frame, we attempt to build frame with it and return.
+      // Other cases, create a composite buffer to manage all the buffers.
+      if (buffers.isEmpty()) {
+        return null;
+      } else if (buffers.getFirst().readableBytes() >= frameRemainingBytes) {
+        // Reset buf and size for next frame.
+        frameBuf = null;
+        nextFrameSize = UNKNOWN_FRAME_SIZE;
+        return nextBufferForFrame(frameRemainingBytes);
+      } else {
+        frameBuf = buffers.getFirst().alloc().compositeBuffer(Integer.MAX_VALUE);
+      }
     }
 
-    // Otherwise, create a composite buffer.
-    CompositeByteBuf frame = buffers.getFirst().alloc().compositeBuffer(Integer.MAX_VALUE);
-    while (remaining > 0) {
-      ByteBuf next = nextBufferForFrame(remaining);
-      remaining -= next.readableBytes();
-      frame.addComponent(next).writerIndex(frame.writerIndex() + next.readableBytes());
+    while (frameRemainingBytes > 0 && !buffers.isEmpty()) {
+      ByteBuf next = nextBufferForFrame(frameRemainingBytes);
+      frameRemainingBytes -= next.readableBytes();
```

Comment:
Just checking here: is there any chance that `readableBytes()` changes value between these two calls?
And does checking `writerIndex()` _after_ `addComponent` give the desired result?

Reply:
If no exception occurred in `addComponent`, these two calls are the same, and `writerIndex` after `addComponent` does give desired result because `addComponent` here does not increase writerIndex.
Actually, these code are almost the same with master branch. So I think it's ok.